SWIPs/swip-honey_to_money.md

```diff
@@ -42,10 +42,10 @@ In the current implementation, honey prices are hardcoded. When this price in co
 
```

Comment:
I guess this timestamp would be added to the signed content of the Cheque?

Reply:
This seems not sufficiently handled. If a time entry is passed, what does the node need to do? Does it need to query the oracle for that specific time? If yes, then the oracle needs also a function `HoneyToMonetAt`, or the standard function needs to take an optional date as parameter. This should be specified.

In any case this also adds new error conditions. What if I send a time when the price was really low? This means that the receiver would also have to check the time parameter to decide if it is even valid, e.g. allow not older than X blocks for example, or something like that. This needs closer examination.

Reply:
In the first honey/money oracle design, this payload was included. Later iterations got rid of this payload, especially because of the added complexity which you correctly uncover here. 

I would vote myself for getting rid of the payload, and naively assume that the oracle is queried in the same block as the transaction is received. This will cause accounting dust to accumulate, but when the honey/currency exchange rate is stochastic (not predictable, which it should be), then the accounting dust will be zero on the long-term. Adding @homotopycolimit and @vojtech here to get their opinions.

Reply:
Hi @Eknir, can you stop using my user name(@vojtech)? Your developer's username is @vojtechsimetka. Thanks

Reply:
Some thoughts:
While the smart contract of the oracle does not need a 'HoneyToMoneyAt' function itself, the swarm module querying the oracle data must be able to query prices in the recent past. I assume this is possible for any full node to answer and does not require an archive node. Yes?

> What if I send a time when the price was really low?

We expect the oracle prices to change neither rapidly nor frequently. So the games you are hinting at are somewhat limited in scope. Suppose the prices are updated once every 24 hours... then at the moment the price increases, a node may quickly send a cheque backdated to 'a few minutes ago' to lock in the old price. This is not a fatal problem. All nodes should be configured by default to accept payments like this within a reasonable timeframe (say 5 minutes) similar to how bitcoin nodes accept any new blocks they see if the timestamp falls within N minutes of their expected time... and only if the timestamp is way off is the block rejected.

Nodes could also always settle their debts before the price changes and the most you can get out of this game is to download a lot of data for five minutes after midnight and pay the price from before midnight. I think that is ok.
Besides, in all our SWAP channels we bank on the discipline of repeated dealings. It is not about every individual payment, but the long term. If you try to backdate your cheques too much, you risk disconnection (just as you do when you don't pay).


I would try to limit surprises as much as possible. Let the oracle publish each price with a 'valid until' timestamp and also declare a time (before the 'valid until' expires) when the new price is published. 
[Simple example: new prices published every evening at 18:00 and go into effect at midnight. In this scenario the 'payload' needed is just the date, and nodes accept yesterdays date until a few minutes after midnight (but do not publish by how many minutes). Frequent late payments can be used to lower a nodes reputation in a future yet-to-be-designed peer reputation scoring system.].

Reply:
@holisticode We removed the requirement to sent timestamp with the payment because:

1. With some payment mechanisms you can't actually send this information
2. In general the time between sending and receiving payment is negligible so it is reasonable to assume the price did not change in this time.
3. As you correctly pointed out, adding timestamp means a lot new conditions to handle (what if I sand older timestamp? How old is acceptable?...)

That being said, the SWIP does not forbid sending any additional information. It just suggests the minimal set needed.

Reply:
> With some payment mechanisms you can't actually send this information

What payment mechanism cannot send any metadata/payment reference at all along with a payment?

> In general the time between sending and receiving payment is negligible

until it isn't. Furthermore you cannot assume that you see the latest block at the same time as your peers do. You have to allow at least some buffer.

> (what if I sand older timestamp? How old is acceptable?...)

If you send one that is too old, I may hold that against you and reduce my opinion/rating of you. Ultimately I may disconnect you just like I might do if you do not pay at all.

> How old is acceptable? 

I would set a short time in minutes - say 5. So let's say that I see a payment which references a latest block that is less than 5 minutes old, I should accept it as valid. As long as the internet keeps functioning as it does today, it is very unlikely that we will ever see this threshold breached by honest nodes.

Reply:
> What payment mechanism cannot send any metadata/payment reference at all along with a payment?

For example normal transaction on Ethereum.

Reply:
sure you can send data along with a payment in ethereum.

Reply:
I don't understand, the SWIP says "To avoid this situation, we propose that nodes include a payload with the transaction, specifying the time at which the oracle was queried", while @vojtechsimetka says "We removed the requirement to sent timestamp with the payment because:".

So what is being sent exactly?

Reply:
Discussion started since the first interation described an optional payload, which was removed in later iterations and added again (yay! Distributed document editing :D). Let's converge towards not sending any payload initially, but describing the optionality of sending a payload - and leaving any further details open for now. At least, unless somebody has a strong opinion against? @vojtechsimetka @homotopycolimit